### PR TITLE
[Merged by Bors] - refactor(Topology/Category): add CompHausLike.Basic

### PR DESCRIPTION
**Pull request by dagurtomas**
This is the first part of the refactor of `CompHaus` and friends (adding the new API which will eventually be used to redefine  the categories of compact Hausdorff spaces, see #12930)

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Topology/Category/CompHausLike/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+  /-- The underlying topological space of an object of `CompHausLike P`. -/
+  toTop : TopCat
+  /-- The underlying topological space is compact. -/
+  [is_compact : CompactSpace toTop]
+  /-- The underlying topological space is T2. -/
+  [is_hausdorff : T2Space toTop]
+  /-- The underlying topological space satisfies P. -/
+  prop : P toTop
+
+-- Porting note (#10754): Adding instance
```
Comment (riccardobrasca) [inline Mathlib/Topology/Category/CompHausLike/Basic.lean:34]: How can we have a porting note on new material?

Comment (dagurtomas) [inline Mathlib/Topology/Category/CompHausLike/Basic.lean:34]:
```suggestion
```

File: Mathlib/Topology/Category/CompHausLike/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+attribute [instance] CompHausLike.is_compact
+-- Porting note (#10754): Adding instance
```
Comment (dagurtomas) [inline Mathlib/Topology/Category/CompHausLike/Basic.lean:36]:
```suggestion
```

File: Mathlib/Topology/Category/CompHausLike/Basic.lean [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+@[simp]
+theorem coe_id (X : CompHausLike P) : (𝟙 ((forget (CompHausLike P)).obj X)) = id :=
+  rfl
+
+@[simp]
+theorem coe_comp {X Y Z : CompHausLike P} (f : X ⟶ Y) (g : Y ⟶ Z) :
+    ((forget (CompHausLike P)).map f ≫ (forget (CompHausLike P)).map g) = g ∘ f :=
+  rfl
+
+-- "Porting" note (#10754): Adding instance
```
Comment (dagurtomas) [inline Mathlib/Topology/Category/CompHausLike/Basic.lean:83]:
```suggestion
-- Note (#10754): Lean does not see through the forgetful functor here
```

File: Mathlib/Topology/Category/CompHausLike/Basic.lean [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+instance (X : CompHausLike.{u} P) : TopologicalSpace ((forget (CompHausLike P)).obj X) :=
+  show TopologicalSpace X.toTop from inferInstance
+
+-- "Porting" note (#10754): Adding instance
```
Comment (dagurtomas) [inline Mathlib/Topology/Category/CompHausLike/Basic.lean:87]:
```suggestion
-- Note (#10754): Lean does not see through the forgetful functor here
```

File: Mathlib/Topology/Category/CompHausLike/Basic.lean [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+instance (X : CompHausLike.{u} P) : CompactSpace ((forget (CompHausLike P)).obj X) :=
+  show CompactSpace X.toTop from inferInstance
+
+-- "Porting" note (#10754): Adding instance
```
Comment (dagurtomas) [inline Mathlib/Topology/Category/CompHausLike/Basic.lean:91]:
```suggestion
-- Note (#10754): Lean does not see through the forgetful functor here
```

File: Mathlib/Topology/Category/CompHausLike/Basic.lean [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+  [is_compact : CompactSpace toTop]
+  /-- The underlying topological space is T2. -/
+  [is_hausdorff : T2Space toTop]
+  /-- The underlying topological space satisfies P. -/
+  prop : P toTop
+
+attribute [instance] CompHausLike.is_compact
+attribute [instance] CompHausLike.is_hausdorff
+
+namespace CompHausLike
```
Comment (joelriou) [inline Mathlib/Topology/Category/CompHausLike/Basic.lean:37]:
```suggestion
namespace CompHausLike

attribute [instance] is_compact is_hausdorff

```

File: Mathlib/Topology/Category/CompHausLike/Basic.lean [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+  rfl
+
+@[simp]
+theorem coe_comp {X Y Z : CompHausLike P} (f : X ⟶ Y) (g : Y ⟶ Z) :
+    ((forget (CompHausLike P)).map f ≫ (forget (CompHausLike P)).map g) = g ∘ f :=
+  rfl
+
+-- Note (#10754): Lean does not see through the forgetful functor here
+instance (X : CompHausLike.{u} P) : TopologicalSpace ((forget (CompHausLike P)).obj X) :=
+  show TopologicalSpace X.toTop from inferInstance
```
Comment (joelriou) [inline Mathlib/Topology/Category/CompHausLike/Basic.lean:83]: A slightly better syntax is:
```suggestion
  inferInstanceAs (TopologicalSpace X.toTop)
```

File: Mathlib/Topology/Category/CompHausLike/Basic.lean [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+
+instance concreteCategory : ConcreteCategory (CompHausLike P) :=
+  InducedCategory.concreteCategory _
+
+instance hasForget₂ : HasForget₂ (CompHausLike P) TopCat :=
+  InducedCategory.hasForget₂ _
+
+variable (X : Type u) [TopologicalSpace X] [CompactSpace X] [T2Space X]
+
+/-- This wraps the predicate `P` in a typeclass. -/
```
Comment (joelriou) [inline Mathlib/Topology/Category/CompHausLike/Basic.lean:53]:
```suggestion
/-- This wraps the predicate `P : TopCat → Prop` in a typeclass. -/
```

File: Mathlib/Topology/Category/CompHausLike/Basic.lean [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+  TopCat.homeoOfIso <| (compHausLikeToTop P).mapIso f
+
+/-- The equivalence between isomorphisms in `CompHaus` and homeomorphisms
+of topological spaces. -/
+@[simps]
+def isoEquivHomeo {X Y : CompHausLike.{u} P} : (X ≅ Y) ≃ (X ≃ₜ Y) where
+  toFun := homeoOfIso
+  invFun := isoOfHomeo
+  left_inv _ := rfl
+  right_inv _ := rfl
```
Comment (joelriou) [inline Mathlib/Topology/Category/CompHausLike/Basic.lean:205]:
```suggestion
  right_inv _ := rfl

end CompHausLike

```

Comment (dagurtomas) [reply Mathlib/Topology/Category/CompHausLike/Basic.lean:205]: Thanks, I always forget this!

File: Mathlib/Topology/Category/CompHausLike/Basic.lean [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+faithful. -/
+def fullyFaithfulToCompHausLike {P P' : TopCat → Prop}
+    (h : ∀ (X : CompHausLike P), P X.toTop → P' X.toTop) : (toCompHausLike h).FullyFaithful :=
+  fullyFaithfulInducedFunctor _
+
+instance {P P' : TopCat → Prop} (h : ∀ (X : CompHausLike P), P X.toTop → P' X.toTop) :
+    (toCompHausLike h).Full := (fullyFaithfulToCompHausLike h).full
+
+instance {P P' : TopCat → Prop} (h : ∀ (X : CompHausLike P), P X.toTop → P' X.toTop) :
+    (toCompHausLike h).Faithful := (fullyFaithfulToCompHausLike h).faithful
```
Comment (joelriou) [inline Mathlib/Topology/Category/CompHausLike/Basic.lean:113]:
```suggestion
section

variable {P P' : TopCat → Prop} (h : ∀ (X : CompHausLike P), P X.toTop → P' X.toTop)

/-- If `P` imples `P'`, then the functor from `CompHausLike P` to `CompHausLike P'` is fully
faithful. -/
def fullyFaithfulToCompHausLike : (toCompHausLike h).FullyFaithful :=
  fullyFaithfulInducedFunctor _

instance : (toCompHausLike h).Full := (fullyFaithfulToCompHausLike h).full

instance : (toCompHausLike h).Faithful := (fullyFaithfulToCompHausLike h).faithful

end
```

**Comment by joelriou** [comment]
Thanks!

bors d+

**Comment by dagurtomas** [comment]
Thanks for the reviews!

bors merge

**Pull request closed without merging**